### PR TITLE
Add jaeger for tracing.

### PR DESCRIPTION
**Pull request by rghetia**

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Should we add the ocagent exporter and deploy the agent instead? That way the whole demo can be run with any exporter.

**Comment by rghetia** [comment]
> Should we add the ocagent exporter and deploy the agent instead? That way the whole demo can be run with any exporter.

That would be much better. As soon as OC Agent (Server) is ready I can switch it over.

**Comment by bogdandrutu** [comment]
We should have a first working version by the end of the week.

**Comment by bogdandrutu** [comment]
Java ocagent exporter may take more to be released

File: src/productcatalogservice/server.go [hunk 1]
```diff
@@ -87,6 +104,14 @@ func initStats(exporter *stackdriver.Exporter) {
 }
 
 func initTracing() {
+	// This is a demo app with low QPS. trace.AlwaysSample() is used here
+	// to make sure traces are available of observation and analysis.
```
Comment (rghetia) [inline src/productcatalogservice/server.go]: s/of/for/

File: src/frontend/main.go [hunk 2]
```diff
@@ -189,6 +206,14 @@ func initStats(log logrus.FieldLogger) {
 }
 
 func initTracing(log logrus.FieldLogger) {
+	// This is a demo app with low QPS. trace.AlwaysSample() is used here
+	// to make sure traces are available of observation and analysis.
```
Comment (rghetia) [inline src/frontend/main.go]: s/of/for/

File: src/checkoutservice/main.go [hunk 3]
```diff
@@ -135,6 +152,14 @@ func initStats() {
 }
 
 func initTracing() {
+	// This is a demo app with low QPS. trace.AlwaysSample() is used here
+	// to make sure traces are available of observation and analysis.
```
Comment (rghetia) [inline src/checkoutservice/main.go]: s/of/for/

File: src/shippingservice/main.go [hunk 4]
```diff
@@ -122,6 +139,14 @@ func initStats(exporter *stackdriver.Exporter) {
 }
 
 func initTracing() {
+	// This is a demo app with low QPS. trace.AlwaysSample() is used here
+	// to make sure traces are available of observation and analysis.
```
Comment (rghetia) [inline src/shippingservice/main.go]: ditto.

File: src/frontend/Gopkg.lock [hunk 5]
```diff
@@ -16,26 +16,32 @@
...
+  name = "git.apache.org/thrift.git"
+  packages = ["lib/go/thrift"]
+  pruneopts = "UT"
+  revision = "014f53f6582fbae8146ae291d471382016091a06"
+  source = "github.com/apache/thrift"
+
+[[projects]]
+  branch = "master"
+  digest = "1:6cbe7676244a1429f4c22601f799d377a70449469ef636f91d992d719b559ff3"
   name = "github.com/GoogleCloudPlatform/microservices-demo"
```
Comment (sebright) [inline src/frontend/Gopkg.lock]: This isn't part of the PR, but I think this dependency on the Github version could be removed.

File: src/productcatalogservice/server.go [hunk 6]
```diff
@@ -77,6 +78,22 @@ func run(port int) string {
...
 }
 
+func initJaegerTracing() {
+
+	// Register the Jaeger exporter to be able to retrieve
+	// the collected spans.
+	exporter, err := jaeger.NewExporter(jaeger.Options{
+		Endpoint: "http://jaeger:14268",
+		Process: jaeger.Process{
+			ServiceName: "frontend",
```
Comment (sebright) [inline src/productcatalogservice/server.go]: Is this a copy/paste error?

File: src/shippingservice/main.go [hunk 7]
```diff
@@ -112,6 +113,22 @@ func (s *server) ShipOrder(ctx context.Context, in *pb.ShipOrderRequest) (*pb.Sh
...
 }
 
+func initJaegerTracing() {
+
+	// Register the Jaeger exporter to be able to retrieve
+	// the collected spans.
+	exporter, err := jaeger.NewExporter(jaeger.Options{
+		Endpoint: "http://jaeger:14268",
+		Process: jaeger.Process{
+			ServiceName: "frontend",
```
Comment (sebright) [inline src/shippingservice/main.go]: ditto

**Pull request merged**